### PR TITLE
refactor: instrument using new telemetry APIs

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

upstream: https://github.com/awslabs/smithy-kotlin/pull/866

## Description of changes

* Refactor instrumentation of the SDK to use the new telemetry APIs
* Extend instrumentation for AWS specific attributes (e.g. `request_id`)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/AbstractAwsSdkClientFactory.kt [hunk 1]
```diff
@@ -48,14 +49,15 @@ public abstract class AbstractAwsSdkClientFactory<
...
-        val tracer = if (builder is TracingClientConfig.Builder) {
-            if (builder.tracer == null) builder.tracer = defaultTracer(builder.config.clientName)
-            builder.tracer!!
+        // FIXME - use a default telemetry provider that at least wires up SLF4j logging
+        val telemetryProvider = if (builder is TelemetryConfig.Builder) {
+            builder.telemetryProvider ?: TelemetryProvider.None
         } else {
-            defaultTracer(builder.config.clientName)
+            TelemetryProvider.None
         }
```
Comment (ianbotsf) [inline aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/AbstractAwsSdkClientFactory.kt:57]: **Nit**:

```kotlin
val telemetryProvider = (builder as? TelemetryConfig.Builder)?.telemetryProvider ?: TelemetryProvider.None
```

File: aws-runtime/aws-http/common/src/aws/sdk/kotlin/runtime/http/interceptors/AwsSpanInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    override suspend fun modifyBeforeCompletion(context: ResponseInterceptorContext<Any, Any, HttpRequest?, HttpResponse?>): Result<Any> {
+        // ensure the overall operation span gets request ID attributes
+        val httpResp = context.protocolResponse
+        val span = coroutineContext.traceSpan
+        if (httpResp != null && span != null) {
+            setAwsRequestIdAttrs(span, httpResp)
+        }
+        return super.modifyBeforeCompletion(context)
+    }
```
Comment (ianbotsf) [inline aws-runtime/aws-http/common/src/aws/sdk/kotlin/runtime/http/interceptors/AwsSpanInterceptor.kt:41]: **Suggestion**: The null checks are common to both methods and could be moved into `setAwsRequestIdAttrs`.

**Review by ianbotsf — APPROVED** [review]

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/protocols/middleware/AwsSpanInterceptor.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
+import software.amazon.smithy.kotlin.codegen.model.buildSymbol
+import software.amazon.smithy.kotlin.codegen.model.namespace
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolMiddleware
+import software.amazon.smithy.model.shapes.OperationShape
+
+private const val AWS_SPAN_INTERCEPTOR_NAME = "AwsSpanInterceptor"
+
+class AwsSpanInterceptor : ProtocolMiddleware {
```
Comment (lauzadis) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/protocols/middleware/AwsSpanInterceptor.kt:18]: nit: I would recommend changing the name of this to `AwsSpanInterceptorMiddleware` to differentiate from the actual interceptor class

**Review by lauzadis — APPROVED** [review]

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/profile/AwsConfigLoader.kt [hunk 4]
```diff
@@ -30,20 +30,21 @@ public suspend fun loadAwsSharedConfig(
     platform: PlatformProvider,
     profileNameOverride: String? = null,
 ): AwsSharedConfig =
-    coroutineContext.withChildTraceSpan("loadAwsSharedConfig") {
+    withSpan("AwsSharedConfig", "loadAwsSharedConfig") {
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/profile/AwsConfigLoader.kt:33]: Define correct. There is no "component" here just a free function so it's kind of whatever we want to put. The span name is the name of the function `loadAwsSharedConfig`.

**Pull request merged**